wavelength-plugins/lottery.js

```diff
@@ -0,0 +1,121 @@
+		this.lottoNumber = room.lottoNumber;
+		this.costToJoin = 3;
+		this.state = "signups";
+		this.playerCount = 0;
+		this.room.add(`|uhtml|lottery-${this.lottoNumber}|<div class="broadcast-blue"><p style="font-size: 14pt; text-align: center">A new <strong>Lottery drawing</strong> is starting!</p><p style="font-size: 9pt; text-align: center"><button name="send" value="/lotto join">Join</button><br /><strong>DISCLAIMER: Joining costs ${this.costToJoin} ${currencyPlural}!!!!</strong></p></div>`, true);
+		this.timer = setTimeout(() => {
+			if (this.playerCount < 2) {
+				this.room.add('|uhtmlchange|lottery-' + this.lottoNumber + '|<div class="broadcast-red"><p style="text-align: center; font-size: 14pt>This Lottery drawing has ended due to lack of users.</p></div>');
+				return this.end();
+			}
+			this.drawWinner();
+		}, 1000 * 60 * 60 * 24);
+	}
+
+	onConnect(user, connection, room) {
+		if (this.state === "signups") {
+			user.sendTo(this.room, '|uhtml|lottery-' + this.lottoNumber + '|<div class="broadcast-blue"><p style="text-align: center; font-size: 14pt>A Lottery Drawing has started looking for players!<hr /><br />For the price of 3 ' + currencyPlural + ', you can earn 5 ' + currencyPlural + ' plus one ' + currencyName + ' per user who joins.</p><br /><button name="send" value="/lottery join">Click here to join the Lottery</button></div>');
+		} else {
+			user.sendTo(this.room, '|uhtmlchange|lottery-' + this.lottoNumber + '|<div class="broadcast-green"><p style="text-align: center; font-size: 14pt>This Lottery drawing is about to announce a winner!!!');
+		}
+	}
+
+	drawWinner(user) {
+		let basePrize = 5;
+		let lottoPrize = basePrize + this.playerCount + this.costToJoin;
+		let winner = toId(user.name);
+		this.room.add(`|html|<div class="infobox"><center><strong>Congratulations</strong> ${WL.nameColor(winner, true)}!!! You have won the reward of ${lottoPrize} ${currencyPlural}</center></div>`);
+		Economy.writeMoney(winner, lottoPrize);
+		Economy.logTransaction(`${winner} has won the Lottery prize of ${lottoPrize} ${currencyPlural}`);
+		this.state = "drawing";
+		this.end();
+	}
+
+	//TODO: Deny users from joining, if they already have joined
+	joinLottery(user) {
+		Economy.readMoney(user.userid, money => {
+			if (money < this.costToJoin) {
+				user.sendTo(this.room, 'You have been removed from this Lottery drawing, as you do not have enough ' + currencyPlural + ' to join.');
+				return;
+			}
+			Economy.writeMoney(user.userid, -this.costToJoin, () => {
+				Economy.readMoney(user.userid, money => {
+					Economy.logTransaction(user.userid + " entered a Lottery drawing for " + this.costToJoin + " " + currencyPlural + ".");
+				});
+			});
+		});
+		this.playerCount++;
+	}
+
+	//TODO: Add a check that the user must be already in the Lottery drawing before able to leave
+	leaveLottery(user) {
+		user.sendTo(this.room, '|html|' + WL.nameColor(user.name, true) + ' has left the game!');
```

Comment:
Since only the user who is leaving can see it, why not just say `You have left the lottery and been refunded ${cost}`